### PR TITLE
Aarch64 Use immediate forms of branch-and-link if possible

### PR DESCRIPTION
**Pull request by jim-saxman**
Summary:
When emiting a call, try and use a pc-relative bl instruction if
possible.

### CODE REVIEW

**Comment by jim-saxman** [comment]
Hi @mxw, @swalk-cavium, can you please review?

**Comment by ottoni** [comment]
I think this should be done in the relocator instead, because code is typically not emitted in its final place.

**Comment by jim-saxman** [comment]
Hi @ottoni, thanks for the quick review. I was targeting the call's emitted in the unique-stubs*, which are not relocated, correct?

**Comment by ottoni** [comment]
@jim-saxman, I see. Yeah, I think the unique stubs are not relocated.

One concern about this is that, for code that is relocated, I'm not sure the relocator currently knows how to turn the call back into a BLR in case the offset doesn't fit into the immediate after the code is relocated. Can you please check this and, if needed, add such support to the relocator?

**Comment by jim-saxman** [comment]
Hi @ottoni, The relevant code in relocation-arm.cpp is around the use of IsUncondBranchImm(). Good thing you asked me to check, since this function returns true for both B (imm) and BL (imm). The only difference in these two instructions' encoding is the 31st bit.

In relocatePCRelative() if IsUncondBranchImm() is true and is_int26(imm) is false the relocator will always use a BR instruction. I have a fix which will output a BLR instruction if that's what is called for. I will post it after the unit testing completes.

File: hphp/runtime/vm/jit/relocation-arm.cpp [hunk 1]
```diff
@@ -610,6 +612,7 @@ bool relocatePCRelative(Env& env, TCA srcAddr, TCA destAddr,
         isRelative = false;
       }
     } else if (src->IsUncondBranchImm()) {
+      // Handle both B and BL forms of UncondBranchImm.
       imm >>= kInstructionSizeLog2;
       if (!is_int26(imm) || env.far.count(src)) {
         env.destBlock.setFrontier(destAddr);
```
Comment (jim-saxman) [inline hphp/runtime/vm/jit/relocation-arm.cpp]: I couldn't exercise this code as is during unit testing or OSS/Mediawiki. However, I forced it by using `if (!is_int6(imm) || env.far.count(src)) {` on the previous line.

**Pull request closed without merging**